### PR TITLE
[13] [UI] Forgot Password screen

### PR DESCRIPTION
**Pull request by markgravity**
https://github.com/markgravity/flutter-survey/issues/13

## What happened

Create UI for Forgot Password screen

### CODE REVIEW

**Comment by manh-t** [comment]
@markgravity Please resolve conflicts, then I can continue reviewing 🙏

**Comment by markgravity** [comment]
> @markgravity Please resolve conflicts, then I can continue reviewing 🙏

Sorry, done it! 👌

File: lib/modules/forgot_password/forgot_password_view.dart [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  void _dismissResetSuccessfullyNotification() {
+    _dismissNotificationOperation?.cancel();
+    _notificationAnimationController.reverse();
+  }
+
+  @override
+  void dispose() {
+    _notificationAnimationController.dispose();
+    _dismissNotificationOperation?.cancel();
+    super.dispose();
```
Comment (manh-t) [inline lib/modules/forgot_password/forgot_password_view.dart:205]: need disposing `_emailTextController` I think 😅

**Review by manh-t — COMMENTED** [review]
could you please separate UI and Backend tasks next times? 😅 I would help us process review faster as we focus on a specific part 💪

File: lib/modules/forgot_password/forgot_password_view.dart (hunk 1 above)
Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_view.dart:206]: @manh-t thanks. I updated them! 🙇

File: lib/modules/forgot_password/forgot_password_presenter.dart [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    implements ForgotPasswordViewDelegate, ForgotPasswordInteractorDelegate {
+  @override
+  final emailTextFieldDidChange = BehaviorSubject<String>();
+
+  @override
+  final resetButtonDidTap = BehaviorSubject<String>();
+
+  @override
+  final alertDialogDidClose = BehaviorSubject<void>();
+}
```
Comment (manh-t) [inline lib/modules/forgot_password/forgot_password_presenter.dart:16]: @markgravity did you forget adding `.addTo(disposeBag)` for these vars? 👀

Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_presenter.dart:16]: @manh-t Thanks! I updated it in [here](https://github.com/markgravity/flutter-survey/blob/5b617d3b2fcbc0ede0a43641b6ed78c14f2d2c4a/lib/modules/forgot_password/forgot_password_presenter.dart#L9)

**Review by manh-t — APPROVED** [review]
rest LGTM 🐰

File: lib/modules/forgot_password/components/reset_successfully_notification.dart [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            Assets.images.forgotPasswordBellIcon.svg(),
+            const SizedBox(
+              width: 15.0,
+            ),
+            Column(
+              crossAxisAlignment: CrossAxisAlignment.start,
+              mainAxisSize: MainAxisSize.min,
+              children: const [
+                Text(
+                  "Check your email.",
```
Comment (sleepylee) [inline lib/modules/forgot_password/components/reset_successfully_notification.dart:25]: these are not translated? 🤔

File: lib/modules/forgot_password/components/reset_successfully_notification.dart [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                  style: TextStyle(
+                    color: Colors.white,
+                    fontWeight: FontWeight.bold,
+                    fontSize: 13.0,
+                  ),
+                ),
+                Text(
+                  "We’ve email you instructions to reset 
your password.",
```
Comment (sleepylee) [inline lib/modules/forgot_password/components/reset_successfully_notification.dart:33]: these are not translated? 🤔

File: lib/modules/forgot_password/components/reset_successfully_notification.dart [hunk 5]
```diff
@@ -0,0 +1,46 @@
+part of '../forgot_password_module.dart';
+
+class ResetSuccessfullyNotification extends StatelessWidget {
+  const ResetSuccessfullyNotification({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      color: const Color(0xff252525).withOpacity(0.6),
```
Comment (sleepylee) [inline lib/modules/forgot_password/components/reset_successfully_notification.dart:9]: a common practice for these custom color is to extract some sort of constant for them. Something like the standard `Colors` class:
```
Colors {
static const neroGray = const Color(0xff252525)
static const neroGray60 = const Color(0xff252525).withOpacity(0.6)

static const pinkyPink = const Color (0xff281921)
}

usage: Colors.neroGray60
``` 

This way you can reuse it, and the naming will be more human-readable.

File: lib/modules/forgot_password/forgot_password_view.dart [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+      // Set a timeout to slide out the notification
+      _dismissNotificationOperation?.cancel();
+      _dismissNotificationOperation = CancelableOperation.fromFuture(
+          Future.delayed(
+              ForgotPasswordView
+                  .dismissResetSuccessfullyNotificationDelayDuration, () {
+        _dismissResetSuccessfullyNotification();
+      }));
+    });
+    super.initState();
```
Comment (sleepylee) [inline lib/modules/forgot_password/forgot_password_view.dart:84]: like @manh-t mentioned, there are too many logic/integrate code in this [UI] PR

File: lib/modules/forgot_password/forgot_password_view.dart [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+  late final Animation<Offset> _notificationAnimation = Tween<Offset>(
+    begin: ForgotPasswordView
+        .resetSuccessfullyNotificationSlideTransitionBeginOffset,
+    end: ForgotPasswordView
+        .resetSuccessfullyNotificationSlideTransitionEndOffset,
+  ).animate(_notificationAnimationController);
+
+  final _emailTextController = TextEditingController();
+  final _isResetButtonEnabled = BehaviorSubject<bool>.seeded(false);
```
Comment (sleepylee) [inline lib/modules/forgot_password/forgot_password_view.dart:63]: it seems a bit strange for me to see this in `view` 🤔  shouldn't the state be kept & process in `presenter`?

File: lib/repositories/auth_repository.dart [hunk 8]
```diff
@@ -99,4 +101,10 @@ class AuthRepositoryImpl implements AuthRepository {
     await attempt();
     await fetchUser();
   }
+
+  @override
+  Future<void> resetPassword({required String email}) {
+    final params = AuthResetPasswordParams(email: email);
+    return _authApiService.resetPassword(params: params);
+  }
```
Comment (sleepylee) [inline lib/repositories/auth_repository.dart:109]: these are the backend ticket :\

File: lib/services/api/auth/params/auth_reset_password_params.dart [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+class AuthResetPasswordParams extends ApiParams {
+  factory AuthResetPasswordParams({
+    required String email,
+    String? clientId,
+    String? clientSecret,
+  }) {
+    return AuthResetPasswordParams._(
+      email: email,
+      clientId: clientId ?? Configs.app.api.clientId,
+      clientSecret: clientSecret ?? Configs.app.api.clientSecret,
```
Comment (sleepylee) [inline lib/services/api/auth/params/auth_reset_password_params.dart:12]: the clientId & clientSecret is unchanged for the whole app session, wouldn't it be better to configure it at the higher level as you did with token?

File: lib/modules/forgot_password/forgot_password_view.dart (hunk 6 above)
Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_view.dart:84]: @sleepylee sorry, I till learn to separate UI and Integrate into difference PR

File: lib/modules/forgot_password/forgot_password_view.dart (hunk 7 above)
Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_view.dart:63]: @sleepylee  It likes we both have view and view model in the same file 😁

File: lib/repositories/auth_repository.dart (hunk 8 above)
Comment (markgravity) [reply lib/repositories/auth_repository.dart:109]: @sleepylee  I know I know, I'm still learning in this process 😢

File: lib/modules/forgot_password/forgot_password_view.dart (hunk 7 above)
Comment (sleepylee) [reply lib/modules/forgot_password/forgot_password_view.dart:63]: you can review the idea again later, technically, if you want to make an `Observable` here, maybe you should have changed the function:
`void setResetButton({required bool isEnabled});` into returning an `Observable`? 

In this case, I support you in making the state `_isResetButtonEnabled` reactive. But this approach is a bit cumbersome right now. It let the Presenter make the call `setResetButton(bool)`, then it applies to the local `BehaviorSubject`. 
So how about we keep all the BehaviorSubject in Presenter, including this, expose read-only Observable to here, then you can map it to your enable state handle of this button? Lol in this case you make a function but we recommend the opposite side 😄  you can refactor it later, but let me know if you get the idea or having objection

**Review by sleepylee — APPROVED** [review]

File: lib/modules/forgot_password/forgot_password_view.dart (hunk 7 above)
Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_view.dart:63]: @sleepylee I gonna document what I'm doing in this project also the reason why I'm doing this, so meet you there to solve your concern 😁

**Pull request merged**